Combine Yul documentation sections. (#8124)
ethereum/solidity: docs/yul.rst

and the last two obfuscate control flow. Furthermore, functional statements of
the form ``mul(add(x, y), 7)`` are preferred over pure opcode statements like
``7 y x add mul`` because in the first form, it is much easier to see which
operand is used for which opcode.

Even though it was designed for stack machines, Yul does not expose the complexity of the stack itself.
The programmer or auditor should not have to worry about the stack.

The third goal is achieved by compiling the
higher level constructs to bytecode in a very regular way.
The only non-local operation performed
by the assembler is name lookup of user-defined identifiers (functions, variables, ...)
and cleanup of local variables from the stack.

To avoid confusions between concepts like values and references,
Yul is statically typed. At the same time, there is a default type
(usually the integer word of the target machine) that can always
be omitted to help readability.

To keep the language simple and flexible, Yul does not have
any built-in operations, functions or types in its pure form.
These are added together with their semantics when specifying a dialect of Yul,

> which allows to specialized Yul to the requirements of different

Reviewer: ```suggestion
which allows to specialize Yul to the requirements of different
```

Yul (previously also called JULIA or IULIA) is an intermediate language that can be
compiled to bytecode for different backends.

Support for EVM 1.0, EVM 1.5 and eWASM is planned, and it is designed to
be a usable common denominator of all three
platforms. It can already be used for "inline assembly" inside Solidity
and there is an experimental implementation of the Solidity compiler
that uses Yul as an intermediate language. Yul is a good target for
high-level optimisation stages that can benefit all target platforms equally.

Motivation and High-level Description
=====================================

The design of Yul tries to achieve several goals:

1. Programs written in it should be readable, even if the code is generated by a compiler from Solidity.

> 2. The translation from assembly to bytecode should contain as few "surprises" as possible.

Reviewer: ```suggestion
2. The translation from assembly to bytecode should be as straightforward as possible.
```
Maybe?
Not sure about writing stuff that needs to be quoted :-).

Reviewer: After reading further below maybe a also:
``and ideally verifiable by manual inspection.``? Or something like that... now I see the point with "few surprises" :-).

> 3. Control flow should be easy to detect to help in formal verification and optimization.

Reviewer: ```suggestion
3. Control flow should only be controlled using a limited set of simple language constructs to help in formal verification and optimization.
```
Maybe? Not sure about "detecting control flow"...

Author: This was also referring to "no jumps" as in llvm IR.

> 4. Yul should be suitable for whole-program optimization

Reviewer: Missing ``.`` in the end.

In order to achieve the first and third goal, Yul provides high-level constructs

> like ``for`` loops, ``if`` and ``switch`` statements and function calls. It should be possible

Reviewer: ```suggestion
like ``for`` loops, ``if`` and ``switch`` statements and function calls. These should be sufficient for adequately representing the control flow for assembly programs. Therefore, no explicit statements for ``SWAP``, ``DUP``, ``JUMP`` and ``JUMPI`` are provided, because...
```
Maybe?

to write EVM assembly programs that do not make use of explicit ``SWAP``, ``DUP``,
``JUMP`` and ``JUMPI`` statements, because the first two obfuscate the data flow
and the last two obfuscate control flow. Furthermore, functional statements of
the form ``mul(add(x, y), 7)`` are preferred over pure opcode statements like
``7 y x add mul`` because in the first form, it is much easier to see which
operand is used for which opcode.

> Even though it was designed for stack machines, Yul does not have an inherent

Reviewer: ```suggestion
Even though it was designed for stack machines, Yul does not expose/hides the complexity of the stack itself, so that this...
```

concept of a stack, so this is something the programmer or auditor does
not have to worry about.

The second goal is achieved by compiling the
higher level constructs to bytecode in a very regular way.
The only non-local operation performed
by the assembler is name lookup of user-defined identifiers (functions, variables, ...),

> which follow very simple and regular scoping rules and cleanup of local variables from the stack.

Reviewer: I'm a bit lost on the last part. Do you mean ``cleanup of local variables from the stack`` as second ``non-local operation``? Then it needs to be ``operation*s*`` and a comma before the ``and``...

Author: Just removed the "which follow very simple..."

> Using ``switch``, ``for`` and user-defined functions, it is possible to write

Reviewer: This is - with and without my suggestion above - kind of what was said before... To me this paragraph feels rather redundant.

Author: I'll re-read the whole thing again and maybe remove this.

complex code without using ``jump`` or ``jumpi`` manually. This makes it much
easier to analyze the control flow, which allows for improved formal
verification and optimization.

> Yul comes in different dialects. The dialect defines the built-in functions

Reviewer: This feels a bit out of place in "Motivation and High-level Description"...
Maybe add in front ``To accomodate the special requirements of different target platforms, different *dialects* of Yul are defined.``?

    }

It is also possible to implement the same function using a for-loop

> instead of with recursion. Here, ``lt(a, b)`` computes the

Reviewer: ```suggestion
instead of with recursion. Here, ``lt(a, b)`` computes whether ``a`` is less than ``b``.
```

        }
    }

Usage
=====

With the "inline assembly" flavour, Yul can be used as a language setting
for the :ref:`standard-json interface <compiler-api>`:

::

    {
        "language": "Yul",
        "sources": { "input.yul": { "content": "{ sstore(0, 1) }" } },
        "settings": {
            "outputSelection": { "*": { "*": ["*"], "": [ "*" ] } },
            "optimizer": { "enabled": true, "details": { "yul": true } }
        }
    }

And on the command line interface with the ``--strict-assembly`` parameter.

.. warning::

>     Yul is in active development and bytecode generation is fully implemented only for untyped Yul (everything is ``u256``)

Reviewer: ```suggestion
    Yul is in active development and bytecode generation is only fully implemented for untyped Yul (everything is ``u256``)
```

    and with EVM 1.0 as target, :ref:`EVM opcodes <opcodes>` are used as built-in functions.

TODO: Explain how to use the optimizer 

Informal Description of Yul
===========================

In the following, we will talk about each individual aspect
of the Yul language. In examples, we will use the default EVM dialect.

Syntax
------

> Yul parses comments, literals and identifiers in the same way as Solidity, so you can use the

Reviewer: ```suggestion
Yul parses comments, literals and identifiers in the same way as Solidity, so you can e.g. use ``//`` and ``/* */`` to signify comments.
```

usual ``//`` and ``/* */`` comments. There is one exception:

> Identifiers in Yul can contain ``.``.

Reviewer: ```suggestion
Identifiers in Yul can contain dots: ``.``.
```

Yul can specify "objects" that consist of code, data and sub-objects.
Please see `Yul Objects <yul-objects>`_ below for details on that.
In this section, we are concerned with the code part of such an object.
This code part always consists of a curly-braces
delimited block. Most tools support a shorthand notation where
only the code block is provided.

Inside a code block, the following elements can be used
(see the later sections for more details):

 - literals, i.e. ``0x123``, ``42`` or ``"abc"`` (strings up to 32 characters)
 - builtin functions, e.g. ``add(1, mload(0))``
 - variable declarations, e.g. ``let x := 7``, ``let x := add(y, 3)`` or ``let x`` (initial value of 0 is assigned)
 - identifiers (variables), e.g. ``add(3, x)``
 - assignments, e.g. ``x := add(y, 3)``
 - blocks where local variables are scoped inside, e.g. ``{ let x := 3 { let y := add(x, 1) } }``
 - if statements, e.g. ``if lt(a, b) { sstore(0, 1) }``
 - switch statements, e.g. ``switch mload(0) case 0 { revert() } default { mstore(0, 1) }
 - for loops, e.g. ``for { let i := 0} lt(i, 10) { i := add(i, 1) } { mstore(i, 7) }``
 - function definitions, e.g. ``function f(a, b) -> c { c := add(a, b) }```

> Multiple syntactical elements are just joined together with

Reviewer: ```suggestion
Multiple syntactical elements can follow each other simply seperated by whitespce,
```

whitespace, i.e. there is no terminating ``;`` or newline required.

Literals
--------

You can use integer constants in decimal or hexadecimal notation.
When compiling for the EVM, this will be translated into an
appropriate ``PUSHi`` instruction. In the following example,
``3`` and ``2`` are added resulting in 5 and then the
bitwise ``and`` with the string "abc" is computed.
The final value is assigned to a local variable called ``x``.
Strings are stored left-aligned and cannot be longer than 32 bytes.

.. code::

    let x := and("abc", add(3, 2))

Function Calls
--------------

Both built-in and user-defined functions (see below) can be called

> in the same way as already seen in the previous example.

Reviewer: ```suggestion
in the same way as shown in the previous example.
```

If the function returns a single value, it can be directly used
inside an expression again. If it returns multiple values,
they have to be assigned to local variables.

.. code::

    mstore(0x80, add(mload(0x80), 3))
    // Here, the user-defined function `f` returns
    // two values. The definition of the function
    // is missing from the example.
    let x, y := f(1, mload(0))

For built-in functions of the EVM, functional expressions

> have a very direct translation to the stream of opcodes:

Reviewer: ```suggestion
can be directly translated to a stream of opcodes:
```

You just read the expression from right to left to obtain the
opcodes. In the case of the first line in the example, this
is ``PUSH1 3 PUSH1 0x80 MLOAD ADD PUSH1 0x80 MSTORE``.

For calls to user-defined functions, the arguments are also
put on the stack from right to left and this is the order
in which argument lists are evaluated. The return values,
though, are expected on the stack from left to right,
i.e. in this example, ``y`` is on top of the stack and ``x``
is below it.

Variable Declarations
---------------------

You can use the ``let`` keyword to declare variables.
A variable is only visible inside the
``{...}``-block it was defined in. When compiling to the EVM,
a new stack slot is created that is reserved
for the variable and automatically removed again when the end of the block
is reached. You can provide an initial value for the variable.

> If you do not provide a value, it will be zero.

Reviewer: ```suggestion
If you do not provide a value, the variable will be initialized to zero.
```

Since variables are stored on the stack, they do not directly
influence memory or storage, but they can be used as pointers
to memory or storage locations in the built-in functions
``mstore``, ``mload``, ``sstore`` and ``sload``.
Future dialects migh introduce specific types for such pointers.

When a variable is referenced, its current value is copied.
For the EVM, this translates to a ``DUP`` instruction.

.. code::

    {
        let zero := 0
        let v := calldataload(zero)
        {
            let y := add(sload(v), 1)
            v := y
        } // y is "deallocated" here
        sstore(v, zero)
    } // v and zero are "deallocated" here

> Depending on the settings, the compiler can already

Reviewer: ```suggestion
Depending on the optimiser settings, the compiler will/can free the stack slot already after...
```

remove the stack slot after the variable has been used for
the last time, even though it is still in scope.

Assignments
-----------

> Variables can be re-assigned arbitrarily using the

Reviewer: "arbitrarily"? As in contrast to what? If anything, maybe "re-assigned to arbitrary values" or "an arbitrary right-hand side"

Author: In contrast to "only once", but maybe I'll just remove tihs.

``:=`` operator. It is possible to assign multiple

> variables at the same time and the number and types

Reviewer: ```suggestion
variables at the same time. For this, the number and types of the values...
```

of values have to match.
If you want to assign the values returned from a function that has
multiple return parameters, you have to provide multiple variables.

.. code::

    let v := 0
    // re-assign v
    v := 2
    let t := add(v, 2)
    function f() -> a, b { }
    // assign multiple values
    v, t := f()

If
--

The if statement can be used for conditionally executing code.

> There is no "else" part, consider using "switch" (see below) if

Reviewer: ```suggestion
No "else" block can be defined. Consider using "switch" instead (see below) if
```

you need multiple alternatives.

.. code::

    if eq(value, 0) { revert(0, 0) }

The curly braces for the body are required.

Switch
------

You can use a switch statement as an extended version of "if/else".
It takes the value of an expression and compares it to several literal constants.
The branch corresponding to the matching constant is taken. Contrary to the

> error-prone behaviour of some programming languages, control flow does

Reviewer: ```suggestion
Contrary to other programming languages, for safety reasons, control flow generally does not...
```

No reason to diss other languages :-). Or at least enough to do it implicitly :-D.

not continue from one case to the next. There can be a fallback or default
case called ``default`` which is taken if none of the literal constants matches.

.. code::

    {
        let x := 0
        switch calldataload(4)
        case 0 {
            x := calldataload(0x24)
        }
        default {
            x := calldataload(0x44)
        }
        sstore(0, div(x, 2))
    }

> The list of cases does not require curly braces, but the body of a

Reviewer: ```suggestion
The list of cases is not enclosed by curly braces, but the body of a
```
Maybe fine to say "not require", but it's not like you can optionally have them - in fact, I do that wrong *every* time :-D.

Author: Actually we might want to change this.

Author: https://github.com/ethereum/solidity/issues/8142

case does require them.

> Loops

Reviewer: I'm stopping here for today.

    v := 2
    let t := add(v, 2)
    function f() -> a, b { }
    // assign multiple values
    v, t := f()

If
--

The if statement can be used for conditionally executing code.
No "else" block can be defined. Consider using "switch" instead (see below) if
you need multiple alternatives.

.. code::

    if eq(value, 0) { revert(0, 0) }

The curly braces for the body are required.

Switch
------

You can use a switch statement as an extended version of the if statement.
It takes the value of an expression and compares it to several literal constants.
The branch corresponding to the matching constant is taken.
Contrary to other programming languages, for safety reasons, control flow does
not continue from one case to the next. There can be a fallback or default
case called ``default`` which is taken if none of the literal constants matches.

.. code::

    {
        let x := 0
        switch calldataload(4)
        case 0 {
            x := calldataload(0x24)
        }
        default {
            x := calldataload(0x44)
        }
        sstore(0, div(x, 2))
    }

The list of cases is not enclosed by curly braces, but the body of a
case does require them.

Loops
-----

> Yul supports for-style loops which consist of

Reviewer: ```suggestion
Yul supports for-loops which consist of
```
?

a header containing an initializing part, a condition, a post-iteration
part and a body. The condition has to be an expression, while
the other three are blocks. If the initializing part
declares any variables at the top level, the scope of these variables extends to all other
parts of the loop.

The ``break`` and ``continue`` statements can be used in the body to exit the loop
or skip to the post-part, respectively.

The following example computes the sum of an area in memory.

.. code::

    {
        let x := 0
        for { let i := 0 } lt(i, 0x100) { i := add(i, 0x20) } {
            x := add(x, mload(i))
        }
    }

> For loops can also be written so that they behave like while loops:

Reviewer: ```suggestion
For loops can also be used as a replacement for while loops:
```

Simply leave the initialization and post-iteration parts empty.

.. code::

    {
        let x := 0
        let i := 0
        for { } lt(i, 0x100) { } {     // while(i < 0x100)
            x := add(x, mload(i))
            i := add(i, 0x20)
        }
    }

Function Declarations
---------------------

Yul allows the definition of functions. When combined with Solidity, these functions

> are different from the Solidity functions and are only available in the inline assembly part.

Reviewer: ```suggestion
are part of a namespace separate from the one for Solidity functions and are only available in the inline assembly part.
```
By the way: this is really true: we *do* allow shadowing functions, I just checked - maybe that should actually be an issue :-).

Author: Actually this is not what I meant to say here. I wanted to say that they are conceptually different.

Reviewer: I'm not sure there's much value in stating that two distinct things are "different" without saying in what way...

For the EVM, these take their
arguments (and a return PC) from the stack and also put the results onto the
stack. User-defined functions and built-in functions are called in exactly the same way.

Functions can be defined anywhere and are visible in the block they are
declared in. Inside a function, you cannot access local variables
defined outside of that function.

Functions declare parameters and return variables, similar to Solidity.
To return a value, you assign it to the return variable(s).

If you call a function that returns multiple values, you have to assign
them to multiple variables using ``a, b := f(x)`` or ``let a, b := f(x)``.

The ``leave`` statement can be used to exit the current function. It

> works like the ``return`` statement in other languages just that it does

Reviewer: Maybe we should mention the builtin return function here?

not take a value to return, it just exits the functions and the function
will return whatever values are currently assigned to the return variable(s).

The following example implements the power function by square-and-multiply.

.. code::

    {
        function power(base, exponent) -> result {
            switch exponent
            case 0 { result := 1 }
            case 1 { result := base }
            default {
                result := power(mul(base, base), div(exponent, 2))
                switch mod(exponent, 2)
                    case 1 { result := mul(base, result) }
            }
        }
    }

Specification of Yul
====================

> This chapter describes Yul code for formally. Yul code is usually placed inside a Yul object,

Reviewer: ```suggestion
This chapter describes Yul code formally. Yul code is usually placed inside Yul objects,
which are ... in their...
```

The ``continue`` and ``break`` statements can only be used inside loop bodies
and have to be in the same function as the loop (or both have to be at the
top level). The ``continue`` and ``break`` statements cannot be used

> in other parts of a loop, not even when it is scoped inside a second loop's body.

Reviewer: I actually didn't know that :-).

The condition part of the for-loop has to evaluate to exactly one value.

The ``leave`` statement can only be used inside a function.

Functions cannot be defined anywhere inside for loop init blocks.

Literals cannot be larger than the their type. The largest type defined is 256-bit wide.

During assignments and function calls, the types of the respective values have to match.
There is no implicit type conversion. Type conversion in general can only be achieved

> if the dialect proivdes an appropriate built-in function that takes a value of one

Reviewer: ```suggestion
if the dialect provides an appropriate built-in function that takes a value of one
```

    E(G, L, n: DecimalNumber) = G, L, dec(n),
        where dec is the decimal decoding function

.. _opcodes:

EVM Dialect
-----------

> The default dialect of Yul currently is the EVM dialect parameterized

Reviewer: ```suggestion
The default dialect of Yul currently is the EVM dialect for the currently selected version of the EVM.
```
?

with a version of the EVM. The only type available in this dialect
is ``u256``, the 256-bit native type of the Ethereum Virtual Machine.
Since it is the default type of this dialect, it can be omitted.

The following table lists all builtin functions
(depending on the EVM version) and provides a short description of the
semantics of the function / opcode.
This document does not want to be a full description of the Ethereum virtual machine,

> so please refer to a different document if you are interested in the precise semantics.

Reviewer: I would suggest to link the yellow paper, but it's still not up to date, is it? *Is* there actually a good choice for  such a "different document" right now :-D? If so we should in fact link it...
We could also remove this sub-clause...

Author: I do not want to impose the burden of the yellow paper to people.

Opcodes marked with ``-`` do not return a result and all others return exactly one value.
Opcodes marked with ``F``, ``H``, ``B``, ``C`` or ``I`` are present since Frontier, Homestead,
Byzantium, Constantinople or Istanbul, respectively.

In the following, ``mem[a...b)`` signifies the bytes of memory starting at position ``a`` up to
but not including position ``b`` and ``storage[p]`` signifies the storage contents at slot ``p``.

Since Yul manages local variables and control-flow,
opcodes that interfere with these features are not available. This includes

> the ``dup`` and ``swap`` instructions as well as ``jump`` instructions and labels.

Reviewer: It says "includes", but maybe still add ``push``?

+=========================+=====+===+=================================================================+
| stop()                  + `-` | F | stop execution, identical to return(0, 0)                       |
+-------------------------+-----+---+-----------------------------------------------------------------+
| add(x, y)               |     | F | x + y                                                           |
+-------------------------+-----+---+-----------------------------------------------------------------+
| sub(x, y)               |     | F | x - y                                                           |
+-------------------------+-----+---+-----------------------------------------------------------------+
| mul(x, y)               |     | F | x * y                                                           |
+-------------------------+-----+---+-----------------------------------------------------------------+
| div(x, y)               |     | F | x / y or 0 if y == 0                                            |
+-------------------------+-----+---+-----------------------------------------------------------------+

> | sdiv(x, y)              |     | F | x / y, for signed numbers in two's complement                   |

Reviewer: ``y = 0``?

Author: does not fit the line :(

Author: ah no it does!

+-------------------------+-----+---+-----------------------------------------------------------------+

> | mod(x, y)               |     | F | x % y                                                           |

Reviewer: Maybe we should also mention the behaviour for ``y = 0`` here, since it's not particularly well defined in general outside the EVM either.

+-------------------------+-----+---+-----------------------------------------------------------------+
| smod(x, y)              |     | F | x % y, for signed numbers in two's complement                   |
+-------------------------+-----+---+-----------------------------------------------------------------+
| exp(x, y)               |     | F | x to the power of y                                             |
+-------------------------+-----+---+-----------------------------------------------------------------+

> | not(x)                  |     | F | ~x, every bit of x is negated                                   |

Reviewer: ```suggestion
| not(x)                  |     | F | bitwise "not" of x (every bit of x is negated)                                   |
```
would be more consistent with the other bitwise operations below

+-------------------------+-----+---+-----------------------------------------------------------------+
| xor(x, y)               |     | F | bitwise "xor" of x and y                                        |
+-------------------------+-----+---+-----------------------------------------------------------------+
| byte(n, x)              |     | F | nth byte of x, where the most significant byte is the 0th byte  |
+-------------------------+-----+---+-----------------------------------------------------------------+
| shl(x, y)               |     | C | logical shift left y by x bits                                  |
+-------------------------+-----+---+-----------------------------------------------------------------+
| shr(x, y)               |     | C | logical shift right y by x bits                                 |
+-------------------------+-----+---+-----------------------------------------------------------------+
| sar(x, y)               |     | C | signed arithmetic shift right y by x bits                       |
+-------------------------+-----+---+-----------------------------------------------------------------+

> | addmod(x, y, m)         |     | F | (x + y) % m with arbitrary precision arithmetic                 |

Reviewer: For these we could also consider mentioning ``m = 0``.

+-------------------------+-----+---+-----------------------------------------------------------------+
| keccak256(p, n)         |     | F | keccak(mem[p...(p+n)))                                          |
+-------------------------+-----+---+-----------------------------------------------------------------+
| pc()                    |     | F | current position in code                                        |
+-------------------------+-----+---+-----------------------------------------------------------------+
| pop(x)                  | `-` | F | discard value x                                                 |
+-------------------------+-----+---+-----------------------------------------------------------------+
| mload(p)                |     | F | mem[p...(p+32))                                                 |
+-------------------------+-----+---+-----------------------------------------------------------------+
| mstore(p, v)            | `-` | F | mem[p...(p+32)) := v                                            |
+-------------------------+-----+---+-----------------------------------------------------------------+

> | mstore8(p, v)           | `-` | F | mem[p] := v & 0xff (only modifies a single byte)                |

Reviewer: Strictly speaking ``mem[p]`` was not defined... but fine...

.. code::

    // A contract consists of a single object with sub-objects representing

>     // the to be deployed code or other contracts that it can create.

Reviewer: ```suggestion
    // the code to be deployed or other contracts it can create.
```
